integrations: Fix jotform integration to receive files.

Varun-Kolanu opened the pull request:
<!-- Describe your pull request here.-->

This PR adds support for receiving files through JotForm. It includes updated tests and fixtures for question-like fields and file handling.

**Approach to handling files:** When a user uploads files in JotForm and submits the form, the rawRequest field in the webhook request body is a string containing JSON. This JSON includes a field named temp_upload, which contains objects with keys representing the names of file fields in the form (e.g., "q7_fileUploads" if the file field is labeled "File Uploads"). These keys are used to extract the corresponding field names (e.g., fileUploads).

The file URLs are sent in the same fields that were extracted earlier. Finally, all the file URLs are rendered along with their respective names.
 
Fixes #32460.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

![001](https://github.com/user-attachments/assets/06411274-63a7-45e7-9d15-f7396f1704df)



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Niloth-p commented on zerver/webhooks/jotform/tests.py:
```diff
@@ -15,7 +15,13 @@ def test_response(self) -> None:
+* **Report Card**: Report Card Niloth.pdf
+
+
+**Files**: [ID.png](https://www.jotform.com/uploads/UserNiloth/243231271343446/6100762336995170500/ID.png), [Report Card Niloth.pdf](https://www.jotform.com/uploads/UserNiloth/243231271343446/6100762336995170500/Report%20Card%20Niloth.pdf)""".strip()
```
Ideally, I think we'd want to avoid having a separate "Files" field, and just link the file values themselves.
But, since a preview of the images are generated, we could list the image names.

Niloth-p commented on zerver/webhooks/jotform/fixtures/response.multipart:
```diff
@@ -65,7 +65,7 @@ Content-Disposition: form-data; name="customBody"
 --------------------------J1JNT4vbVO4FCC1qLG4HSv
 Content-Disposition: form-data; name="rawRequest"
 
-{"slug":"submit\/243231271343446","jsExecutionTracker":"build-date-1732271172685=>init-started:1732615894703=>validator-called:1732615894706=>validator-mounted-false:1732615894706=>init-complete:1732615894707=>onsubmit-fired:1732615897940=>observerSubmitHandler_received-submit-event:1732615897941=>submit-validation-passed:1732615897942=>observerSubmitHandler_validation-passed-submitting-form:1732 ...
```
If we're updating this fixture (the only fixture for Jotform), we'd also need to update the example screenshot in the [integration doc](https://zulip.com/integrations/doc/jotform).

When I ran the generate command for this fixture, it doesn't load the image properly, probably because of lazy loading of the image, not sure.
If you could get the image to load, and hide the tooltip, (by editing [the generator script](https://github.com/zulip/zulip/blob/main/tools/screenshots/generate-integration-docs-screenshot) as an initial commit) we could use this fixture for the screenshot.
Otherwise, we could create a separate fixture file for the files use case, and a separate test.
So that we don't edit the fixture that's used for the screenshot.

Feel free to duplicate this fixture though, because I plan to be updating the fixture used for the screenshot, to a business use case instead, so variety wouldn't be an issue.

Varun-Kolanu replied:
Actually, while manual testing too, if you open the localhost in a browser where you're not logged in with jotform, the images aren't being rendered. Maybe because those form uploads are confidential (yeah obv :smile: )
So, I too couldn't render the images using the generator script. Though, I have manually took screenshot of the webhook message carefully and uploaded it :)

Niloth-p commented on zerver/webhooks/jotform/fixtures/response.multipart:
```diff
@@ -65,7 +65,7 @@ Content-Disposition: form-data; name="customBody"
 Content-Disposition: form-data; name="rawRequest"
 
-{"slug":"submit\/243231271343446","jsExecutionTracker":"build-date-1732271172685=>init-started:1732615894703=>validator-called:1732615894706=>validator-mounted-false:1732615894706=>init-complete:1732615894707=>onsubmit-fired:1732615897940=>observerSubmitHandler_received-submit-event:1732615897941=>submit-validation-passed:1732615897942=>observerSubmitHandler_validation-passed-submitting-form:1732 ...
+{"slug":"submit\/243231271343446","jsExecutionTracker":"build-date-1732271172685=>init-started:1732615894703=>validator-called:1732615894706=>validator-mounted-false:1732615894706=>init-complete:1732615894707=>onsubmit-fired:1732615897940=>observerSubmitHandler_received-submit-event:1732615897941=>submit-validation-passed:1732615897942=>observerSubmitHandler_validation-passed-submitting-form:1732 ...
```
I think it would be nice to have more than one image and more than one non-image file. And maybe a field with multiple linked files.
I tested the above cases manually, but it would help to have automated tests ensure the same.
I like how you've tried out different file extensions, it would help to do the same to maintain variety.

Niloth-p reviewed (commented):
This is pretty neat! Thanks for the PR, @Varun-Kolanu!
This definitely supports files in the payload, thank you!

But we would want to format the message differently, by linking the file values directly, instead of having a separate section at the end for the files.
We could have the separate section for the images though, because their previews are displayed unlike the files.
This would require a change in the view logic, to identify the files from among the field values, and assign them their respective links.

Also, we can remove the "Adds tests and fixtures" from the commit body, because that's necessary to get the commit merged, so it doesn't require an explicit mention.

Varun-Kolanu commented:
Thanks :smiley:

I have updated the logic to include the file urls in the field itself. There is a slight assumption in mine that, files are always kept at last in the `pretty`. Otherwise, its practically impossible to go with this implementation.

Since, here is no way to identify which are the file fields by seeing the pretty field. Because:

Say we have thought to identify the fields with the help of our extracted file fields. But there's a problem here. If any two original field names are same, for example take Upload Files, pretty will be Upload Files: ...., Upload Files: ..... But the fields we extracted from rawRequest will be uploadFiles and uploadFiles13 (they append random numbers i guess to maintain unique urls for images). So, technically we can't decompress the rawRequest file fields into that of pretty's, which makes it not possible to identify the fields in the pretty.

Varun-Kolanu commented:
@Niloth-p Can you have a look at the updated code once?

Niloth-p commented on static/images/integrations/jotform/001.png:
This screenshot looks far better than the one I had in mind, with the addition of the files.

Could you please create a form that generates a fixture that renders a message like this:
```markdown
**How often do you use the application?** Every day
**How likely are you to recommend it to a friend on a scale of 0-10?** 9
**Feedback**: The new personalized recommendations feature is great!
```
And add some relevant files and images for this use case?

The reason being we want to make the example screenshot look like useful additions to Zulip users' business workflows, which the current tutoring use case doesn't satisfy. So, this fixture should be geared towards getting a good screenshot.

This would be a separate second commit that updates the screenshot use case, building upon the first commit's feature addition (so, the duplication of the fixture and tests can be removed from the first commit).

Niloth-p reviewed (commented):
This update looks great, and works great!
Thanks for the quick update!

The assumptions regarding the fixture are sound, and the view logic seems good to me, except for one bug.

Jotform allows hiding questions from the form, and those get sent too, in `raw_request`, but not in `pretty`. So, counting non-file questions from `raw_request` would start calling `replace_with_url_markdowns` for a later index, offset by the number of hidden questions.

Regarding the previous iteration's review comment about using a fixture with a business use case, it would be great if you could add that as a second commit. Left more details in a comment.

Varun-Kolanu commented:
@Niloth-p I have updated my code to eliminate the bug of hidden files in the form too now. Actually, JotForm is excluding all the fields with values having zero lengths. So, not only hidden fields but all the empty fields will be excluded. Hence, I excluded those empty fields in non-file-fields.

Also, I have updated the screenshot to fit into a valid business case.

Niloth-p commented on static/images/integrations/jotform/001.png:
To generate this screenshot, edit https://github.com/zulip/zulip/blob/main/zerver/lib/integrations.py#L758 to use `screenshot_response.multipart`.

I think we'd like to avoid AI-generated images in our example screenshots, but I'll leave the specifics to the next round of review by Alya Abbott (after Jan 2).
And I'm sure that she'll feel that the questions in the example start off like a feedback survey than move onto gathering issue details (which we wouldn't want to use as a business use case, as other tools like GitHub, Jira are better suited for issue reporting than Jotform).

Varun-Kolanu replied:
Yeah I was searching for a suitable business use case but couldn't find more apt use case. And, I searched for suitable images on google which may not get any copyright strike. But all I could find was images involving real people. So I decided to use AI generated images 😅

I think this could be easily updated after Alya reviews and suggests specific use case ✌️

Niloth-p replied:
I don't think we should wait for her to come up with a use case.

On top of the first 3 questions, I think we can just add one more question with something like - "Upload images of your customized setup to get featured in the Community Showcase!"
And a couple of random images like these.

<details>
<summary>Some free images</summary>

![photo1](https://github.com/user-attachments/assets/ce343d21-ccd9-43af-b8c0-fee935a0150c)
![photo2](https://github.com/user-attachments/assets/36948b4e-94b3-4a63-a4fd-640118245140)

</details>

With 1-2 dummy file names like `Samples.pdf`.

Varun-Kolanu replied:
cool cool, gotcha. I'll update it

Varun-Kolanu replied:
Also, I have got that tooltip in the screenshot for the first time. I added this in the `message-screenshot.js` and it worked fine:

![image](https://github.com/user-attachments/assets/48f3cedf-cdb9-4d8d-8af9-f9a58f5a1104)

(Though I haven't added it in the commit now. If this change helps, I shall add it in the same commit or add a new commit?)

Niloth-p reviewed (commented):
Nice catches! Thank you for adding support for those cases as well.

Regarding empty tables, we don't currently support formatting Input Table values in our message, right.
But, that's not exactly necessary for this PR to incorporate, these are good feature additions as they are.
I think a more important Jotform change would be fixing #32461, if you're interested. That could be a separate PR.
(I just noticed that you noted the same issues in a previous version of your comment)

I appreciate all the nice changes like changing the variable name from using "questions" to "fields", using unquote(), and the apt commit message.

Varun-Kolanu commented on zerver/webhooks/jotform/view.py:
```diff
@@ -11,6 +15,48 @@
+
+
+def field_is_non_file_pretty(field: str, raw_request: dict[str, Any]) -> bool:
+    if field[0] != "q" or len(raw_request[field]) == 0:
```
Yeahh it is looking pretty good! Thanks a lot, I have updated this function in the new update.

alya commented:
In the screenshot, let's change:

- Everyday -> Daily
- Remove the pdf upload (there's no preview, so it's just confusing)

alya commented:
Why is there extra vertical space above the "Upload images" bullet?

Varun-Kolanu commented:
> Why is there extra vertical space above the "Upload images" bullet?

Removing the space was rendering the images in the same line, next to the filename, which was making the UI look messy. Adding a "\n" removed it.

Varun-Kolanu commented:
> In the screenshot, let's change:
> 
>     * Everyday -> Daily
> 
>     * Remove the uploaded pdf -- it's confusing, because there's no preview. I see there's some discussion above about adding it, but we don't need to demonstrate _all_ the functionality here. We should avoid situations where the example looks like there's an error unless you take time to think about it.

Done in the new update.

Varun-Kolanu commented:
Also, #32461 is blocked on this PR, since it needs some of the functions defined here. 
(Therefore, I have worked on it, but didn't make a PR)

laurynmm reviewed (commented):
@Varun-Kolanu - Thanks for all your work on this! The changes overall look good. I had some thoughts on the function and variable naming to make it a bit easier to read/understand without deep understanding/knowledge of the form data that Jotform sends.

Let me know if you have any questions about my comments and ping me once you've updated for that feedback so that I can do a quick pass before marking for integration review!

laurynmm commented on zerver/webhooks/jotform/view.py:
```diff
@@ -11,6 +15,46 @@
+    return ", ".join(url_markdown_list)
+
+
+def field_is_non_file_pretty(field: str, raw_request: dict[str, Any]) -> bool:
```
Sure :+1:

laurynmm commented:
Thanks, looks good! Marking for integration review.

timabbott commented:
Merged, thanks @Varun-Kolanu and @Niloth-p!
